### PR TITLE
Enable use of vector.contract within IREE codegen.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**

### CODE REVIEW

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (ThomasRaoux) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:179]: Is this really the final workgroup size? I'm not sure I follow the relationship between tile size, and workgroup size. I think they should be independent as we don't want to create large workgroups.

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (ThomasRaoux) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:529]: Why the number of subgroups is related to the tile size? It should be only dependent on the workgroup size and subgroupSize.

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (ThomasRaoux) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:346]: Maybe should be called getMinCooperativeSize()? Since the size of the cooperative matrix can be bigger but we use that as the smallest granularity?

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (hanhanW) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:53]: Do we need to check/assert the size of attr.getValues is as same as `3` (use workgropuSize.size() would be better).

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (hanhanW) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:47]: According to the comment, if we expect the argument being a function, we should use funcOp here?

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (hanhanW) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:48]: s/operation/op

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (hanhanW) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:86]: can we use `e = std::min(2u, sizes.size())`

and maybe replace `i != e` with `i < e`? I think we typically use `i != e` when it's an iterator.

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (hanhanW) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:173]: use two `/` for comment

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (hanhanW) [reply iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:179]: I think the workgroup size is

1) taking tileSizes from beginning until hit a non-parallel loop. In this case, BatchMatmul is (1, M, N) and Matmul is {M, N}
2) reverse the 1)'s result, BatchMatmul is {N, M, 1} and Matmul is {M, N}/
3) Truncate to size 3 and pad with value `1`.

All the workgroup sizes are {N, M, 1}.

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (hanhanW) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:172]: According to the comment, I'm expecting

tileSizeN = std::ceil(std::sqrt(maxWorkgropuSize));

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (hanhanW) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:345]: s/querried/queried

**Comment by MaheshRavishankar** [comment]
THanks @hanhanW and @ThomasRaoux for the review. After discussion with @ThomasRaoux it turns out the tile size calculation used here doesnt work for the case where we want to use Cooperative Matrix Multiply instructions. This needs a fix to how number of workgroups are computed for an entry point kernel. Will address that first, and then rebase this PR on top of that.

**Comment by MaheshRavishankar** [comment]
@ThomasRaoux @hanhanW @antiagainst This PR is now rebased on top PR #3095 and is ready to be reviewed

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.h
Comment (antiagainst) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.h:73]: Use an empty line to separate with the following

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.h
Comment (antiagainst) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.h:74]: ... the tile sizes, the workgroup size, etc.. We call these parameters together as the "launch configuration". This class...

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.h
Comment (antiagainst) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.h:80]: .. by the code generators,

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.h
Comment (antiagainst) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.h:86]: Document for this please

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.h
Comment (antiagainst) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.h:89]: Gets  to be consistent with the rest.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.h
Comment (antiagainst) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.h:73]: Document for this. What the outer vector is for? Inner vector?

File: iree/compiler/Conversion/LinalgToSPIRV/Passes.cpp
Comment (antiagainst) [inline iree/compiler/Conversion/LinalgToSPIRV/Passes.cpp:57]: ? I thought these ones are for real and the command line ones are for testing purpose?

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (ThomasRaoux) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:227]: Is `dataType.isa<IntegerType>() &&`needed?

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (ThomasRaoux) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:231]: same, is `dataType.isa<FloatType>()` needed?

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (antiagainst) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:110]: Patterns to tile to workgroups ?

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (antiagainst) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:219]: might just use `auto` here? I don't think we care about the detailed function type here?

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp
Comment (antiagainst) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:215]: populateTilingToWorkgroupPatterns is more clear and readable for me. Similarly later we can use populateTilingToSubgroupPatterns.

**Review by antiagainst — COMMENTED** [review]
Nice! Made a first pass and left some comments. Will take another look tomorrow.

You'll also update the doc some time later right?

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (ThomasRaoux) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:272]: I'm not convinced picking the max workgroup size is always going to be the best choice (at least from the experiments I did)?

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (ThomasRaoux) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:278]: I assume we'll want a higher number. Do you plan to handle it in a future PR? Maybe add a TODO?

**Comment by MaheshRavishankar** [comment]
> Nice! Made a first pass and left some comments. Will take another look tomorrow.
> 
> You'll also update the doc some time later right?

Will update the doc yes. Need to update the current way of not using loops as much anyway. Maybe end of month

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:18]: Actually what is there seems better. It is 
- computing tile sizes
- workgroup sizes
- the number of workgroups to use.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:338]: Actually maybe better to check all requirements at the same place given the workgroup size. I added a TODO for now at the end of the `init` function in this file.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:370]: apparently not. Removing this.

File: iree/compiler/Conversion/LinalgToSPIRV/Passes.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/Passes.cpp:57]: This is only for lit-testing. The real options are the `SPIRVCodegenOptions` . Maybe the placement is misleading.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:224]: // is for comments. /// is for doxygen comments ?

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:272]: Agree. For now the default workgroup size is very small. So I am using that to set everything else. That should be flipped and there is more to be done here. Added a TODO.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:278]: Added TODO.

File: iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp
Comment (antiagainst) [inline iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp:181]: s/updated/updates/ to be consistent with others.

**Review by antiagainst — APPROVED** [review]
LGTM to land this to enable future development. It would be nice to wait for @ThomasRaoux's approval too given he has more experience there.

**Review by ThomasRaoux — APPROVED** [review]
Looks good!

**Pull request merged**
